Add a testkit module to hold the TestEvaluator

Baccata opened the pull request:
@lefou opening this as a draft to get an early opinion (if you're against it, it's absolutely fine and I'll just publish it as a third party library from a repo I control) 

the TestEvaluator that the mill codebase uses to test itself is extremely useful for testing some mill modules with reasonably lightweight setup. I have copied it over in some projects a few times now, including [today](https://github.com/indoorvivants/sbt-vcpkg/pull/9#discussion_r890130257). I'd love to have it as a library, as testing modules this way, though not as thorough as a bonafide integration test, is absolutely neat. 

Making it available to downstream libraries and plugins could help expand the ecosystem, in particular when such plugins are built with SBT (which is the harsh reality of the status quo, and makes it hard to use mill-integrationtest). 

If that sounds appealing at all, then I'll wire that in this codebase in a way that would keep the diff to a minimum.

lefou reviewed (commented):
Thanks for bringing this up. Moving the `TestUtil`, `TestEvaluator` and other related classes into a dedicated module (`testutil`) and make it usable outside this repo is IMHO something we definitely want. As we need to think about binary compatibility, we should schedule the merge of such work for the next major release (which is either a `0.11.0` or a `1.0`). Also, we should avoid creating any split packages while doing the refactoring.

I assume the code you've posted until now in this PR is just a placeholder, but for completeness I need to say that it is not acceptable in it's current form because of the duplication of already existing code and also it is not used anywhere or covered by any tests.

Baccata commented:
> I assume the code you've posted until now in this PR is just a placeholder, but for completeness I need to say that it is not acceptable in it's current form because of the duplication of already existing code and also it is not used anywhere or covered by any tests.

That is absolutely correct : the current state is just a **very preliminary** thing I did to get a feel for what you think (in terms of timeframes) and how you'd structure it (hence why it was raise as a draft) 

> we should schedule the merge of such work for the next major release (which is either a 0.11.0 or a 1.0)

Could you share some (even very rough) timeframe you're looking at for 0.11 or 1.0 ? In theory, because this work wouldn't be modifying any published code, it could theoretically be released in the 0.10 lineage in a backward compatible fashion, after careful review to ensure the evolvability. I totally understand if you'd rather wait for a 0.11 though. 

Timeframes aren't a dealbreaker, I think I will get this PR into a mergeable state and then publish its contents from a separate repo to use it ASAP. 

> Also, we should avoid creating any split packages while doing the refactoring. 

Yeah that's a question I had : where should this live in terms of package / module ?

lefou commented:
> > we should schedule the merge of such work for the next major release (which is either a 0.11.0 or a 1.0)
> 
> Could you share some (even very rough) timeframe you're looking at for 0.11 or 1.0 ? 

Mill 0.10 is pretty much stable and out for a while. There is nothing that's us holding back to work on the next iteration except that it needs to be worth it, as it means work in the ecosystem, e.g. external plugins need to create new releases for a new binary platform.

> In theory, because this work wouldn't be modifying any published code, it could theoretically be released in the 0.10 lineage in a backward compatible fashion, after careful review to ensure the evolvability. I totally understand if you'd rather wait for a 0.11 though.

You're absolutely right. I overlooked the fact that it currently resides in a test module. We can just start to publish a `testutil` module as soon as we have it. We can even mark it `@experimental` if we think that the API isn't easy to keep stable for a while.

> > Also, we should avoid creating any split packages while doing the refactoring.
> 
> Yeah that's a question I had : where should this live in terms of package / module ?

I think what you alread did is a good start: `mill.testutil` as top level package and `testutil` as module

Baccata commented:
Awesome then, I'll get on with it and will ping you for an actual review when I have something that I think is ready or close to ready :+1:

Baccata commented:
@lefou I've wired the extracted TestEvaluator into the existing one to minimise changeset. I've also added a test into the new module (now renamed testkit because it seemed more fitting) as a way to showcase how it works. 

I can write some documentation if you'd like, but it'd be worth reviewing the API exposed by the TestEvaluator as it could do with a little bit of polish (in particular in the returned types). 

Also What do I need to avoid CI failing cause my "git history is incomplete" ?

lefou reviewed (commented):
@Baccata Sorry, I need so long for the review. I already glanced at it and think it's good. I want to have a closer look at the diff between `TestEvaluator` and `MillTestkit`. As I have some exams this weekend (voluntary firefighter), I hope I can find the time in the early next week.

Baccata commented:
@lefou, no need to apologise, in particular when you have such a good reason 😄

Hope you rock the exam !

Baccata commented on main/testkit/src/testkit/MillTestkit.scala:
```diff
@@ -0,0 +1,152 @@
+
+import scala.collection.mutable
+
+class MillTestKit(targetDir: os.Path) {
```
> I wonder whether it is a better idea to make targetDir a def inside the class, as we do with other settings like externalOutPath? This would make it more flexible when inherited in an object and would allow some kind of dynamic default.

Sure, I'm cool with that. What would you like the default impl to be ? Something like 

```scala
def targetDir: os.Path = 
  sys.env.get("MILL_TESTKIT_BASEDIR")
       .map(convertToPath) 
       .getOrElse(os.pwd / "out" / "something_with_a_low_collision_risk") ?  
```

Also, is there a built-in function to parse a string into a `RelPath` ?

lefou replied:
Yeah, something like that. Don't know what `convertToPath` is, but `Path.apply(_, os.pwd)` can be used to create a `os.Path` from a string. And `os.rel / aString` can be used to create a `os.RelPath`.

As a fallback, maybe we can create a temp dir with `os.temp.dir`. Either in the system temp dir or under `os.pwd / "target"`?

If we need to read or copy resources in these tests, I nowadays prefer to read them from the classpath. This makes the test independent of the executing build tool and operating system.

Baccata replied:
> If we need to read or copy resources in these tests, I nowadays prefer to read them from the classpath. This makes the test independent of the executing build tool and operating system. 

I think the `target` is ore about where mill's `out` folder gets saved, so it's a "writing" concern more than it is a "reading" one. 

Falling back to `os.temp.dir` makes sense :+1:

lefou replied:
Just a short note, as I saw this code snipped in a second review.

> ```scala
> def targetDir: os.Path = 
>   sys.env.get("MILL_TESTKIT_BASEDIR")
>        .map(convertToPath) 
>        .getOrElse(os.pwd / "out" / "something_with_a_low_collision_risk") ?  
> ```

If I really need dedicated temporary/permanent storage to be shared by more that one target in Mill, I typically end up creating a dedicated persistent target for them. As an example, the target directories we use when we generate IntelliJ IDEA project files, use this.

```scala
// taken from GenIdeaModule.scala
  def ideaCompileOutput: T[PathRef] = T.persistent {
    PathRef(T.dest / "classes")
  }
```

Baccata commented:
> I'd like to keep the API experimental and without compatibility guarantees. I've simple not the capacity to think about all use cases, atm. I could imagine, to add some kind of support to mill-integrationtest plugin, but have no clear idea yet.

Yeah I'm cool with that :+1:

> You mean the version which is derived from last git tag? Depends on your CI. On GitHub, you need can add the fetch-depth: 0 option, which means full history. I remember on Travis, the settings was git: depth: false.

I _think_ I was talking about [the MiMa](https://github.com/com-lihaoyi/mill/runs/6888808411?check_suite_focus=true) job failing, but it makes sense that it would fail because the testkit module has never been published before. I suppose MiMa needs to be disabled on this one, since it won't come with any bincompat guarantees for the time being anyway.

lefou commented on build.sc:
```diff
@@ -419,6 +420,14 @@ object main extends MillModule {
     )
   }
+
+  object testkit extends MillModule {
```
`MillModule` is inheriting `MillApiModule` which also adds `MiMa`. Instead, you could inherit from `MillInternalModule` directly. As a consequence, you need to add the `object test` manually or also inherit `MillAutoTestSetup`.

lefou commented:
> > You mean the version which is derived from last git tag? Depends on your CI. On GitHub, you need can add the fetch-depth: 0 option, which means full history. I remember on Travis, the settings was git: depth: false.
> 
> I _think_ I was talking about [the MiMa](https://github.com/com-lihaoyi/mill/runs/6888808411?check_suite_focus=true) job failing, but it makes sense that it would fail because the testkit module has never been published before. I suppose MiMa needs to be disabled on this one, since it won't come with any bincompat guarantees for the time being anyway.

Oh, I was on the wrong track then. To disable `MiMa`, see my other comment: https://github.com/com-lihaoyi/mill/pull/1893#pullrequestreview-1022936695

lefou commented:
Hi @Baccata, what's the status of this: Do you need anything from my side?

Baccata commented:
Hey @lefou, sorry for the delay, I've been somewhat swamped between the preparation of the next release of smithy4s and some internal things at work. I'm hoping to go back to as soon as smithy4s 0.14 is released (hopefully today or tomorrow)

lefou commented:
> Hey @lefou, sorry for the delay, I've been somewhat swamped between the preparation of the next release of smithy4s and some internal things at work. I'm hoping to go back to as soon as smithy4s 0.14 is released (hopefully today or tomorrow)

Great. Don't stress yourself out, I just wanted to make sure, that I'm not the bottleneck.

Baccata commented:
@lefou, would you kindly do another pass when you have the time ?

lefou reviewed (approved):
I'm going to merge this and I hope, I get a better feeling for it, once I use it. It's still experimental API only used in tests, so there is not much to loose. 

@Baccata Maybe, once you start to use it in an external project, you can also add some guide to the documentation? 

Thank you!

Baccata commented:
> Maybe, once you start to use it in an external project, you can also add some guide to the documentation?

Sure, will do 👍
